src/zkevm_specs/evm_circuit/execution/precompiles/ecrecover.py

```diff
@@ -26,9 +34,25 @@ def ecRecover(instruction: Instruction):
     sig_r: Word = instruction.curr.aux_data[2]
     sig_s: Word = instruction.curr.aux_data[3]
     recovered_addr: FQ = instruction.curr.aux_data[4]
+    rlc_data: PrecompileRlcData = instruction.curr.aux_data[5]
+    keccak_randomness: FQ = instruction.curr.aux_data[6]
 
     is_recovered = FQ(instruction.is_zero(recovered_addr) != FQ(1))
 
+    # Verify input and output
+    input_bytes = bytearray(b"")
+    input_bytes.extend(msg_hash.int_value().to_bytes(32, "little"))
+    input_bytes.extend(sig_v.int_value().to_bytes(32, "little"))
+    input_bytes.extend(sig_r.int_value().to_bytes(32, "little"))
+    input_bytes.extend(sig_s.int_value().to_bytes(32, "little"))
+    input_rlc = RLC(bytes(reversed(input_bytes)), keccak_randomness, n_bytes=128).expr()
+    instruction.constrain_equal(rlc_data.input_rlc, input_rlc)
```

Comment:
Aren't these two values actually the same (I mean not equal, but the same)? Because `rlc_data.input_rlc` is computed when [instantiating the auxiliary data](https://github.com/privacy-scaling-explorations/zkevm-specs/blob/2e5f9263b74ced37ad3d9aea02874ed90bb85bb4/tests/evm/precompiles/test_ecRecover.py#L86) and `input_rlc` is computed accessing the same values in `aux_data`?

Reply:
input_rlc = RLC(msg_hash, sig_v, sig_r, sig_s)

For me, data assignment in testing is like what `assign_exec_step` does at proving time in our circuit. So, yes, it looks the same (but not always the same). 
At proving time, a prover assigns
- msg_hash
- sig_v, sig_r, sig_s
- RLC(msg_hash, sig_v, sig_r, sig_s) (a.k.a `input_rlc` here, this field is to verify data consistency between calls)

In verification logic here, we have to gaurantee msg_hash, sig_v, sig_r and sig_s are the same pairs (sig_v, sig_r and sig_s are coming from the signature of msg_hash). However, a malious signer could sign msg_hash2 and have sig_v2, sig_r2 and sig_s2. It still can pass all the constraints if we don't have calculate input_rlc here. I'm assuming `rlc_data.input_rlc` is coming from previous step (we can pass data around between previous step and next step in our circuit, but it seems not doable in our spec), so that's why `rlc_data.input_rlc` and `input_rlc` look "the same".

Does it make sense to you? Or do you think we need to copy `rlc_data.input_rlc` from copy_table?

Reply:
I think I am missing something and would like to fix my understanding :). My understanding was that we need to ensure that `instruction.curr.aux_data` is the same as used in `callop.py` and that this should be done by checking the RLC, so to have a lookup call into `copy_table` (RLC field)  here in `ecrecover.py`. Maybe this is not the case?

I agree on the assignment part, putting a link to the assignment was probably misleading from my side. What it seems to me here is that we have `aux_data: PrecompileAuxData = instruction.curr.aux_data[0]` and then both values, `aux_data.input_rlc` and `input_rlc`, are computed using `aux_data`. Or is it `aux_data.input_rlc` checked somewhere else?

Reply:
> My understanding was that we need to ensure that `instruction.curr.aux_data` is the same as used in `callop.py` and that this should be done by checking the RLC, so to have a lookup call into `copy_table` (RLC field) here in `ecrecover.py`. Maybe this is not the case?

Yes, that's exactly what I want to do.

> I agree on the assignment part, putting a link to the assignment was probably misleading from my side. What it seems to me here is that we have `aux_data: PrecompileAuxData = instruction.curr.aux_data[0]` and then both values, `aux_data.input_rlc` and `input_rlc`, are computed using `aux_data`. Or is it `aux_data.input_rlc` checked somewhere else?

No, `aux_data.input_rlc` was not check in other place.

In my implementation, I treated `aux_data` as witness inputs, a place I can assign my witnesses. Those data (e.g. sig_r, sig_v...etc) can't be retrieved from stack or memory like what other opcode gagdets implemented. 

I might not explain it very well, but you can check [Scroll's impl.](https://github.com/scroll-tech/zkevm-circuits/blob/57a2dd4c4f8b0443a4cc9ee43c71be649c937ffc/zkevm-circuits/src/evm_circuit/execution/precompiles/ecrecover.rs#L121-L133) The `msg_hash_raw` was converted into rlc formate and compared with `msg_hash_keccak_rlc`. In the assignment, you can see both of them come from the same source, [here](https://github.com/scroll-tech/zkevm-circuits/blob/57a2dd4c4f8b0443a4cc9ee43c71be649c937ffc/zkevm-circuits/src/evm_circuit/execution/precompiles/ecrecover.rs#L372-L380) and [here](https://github.com/scroll-tech/zkevm-circuits/blob/57a2dd4c4f8b0443a4cc9ee43c71be649c937ffc/zkevm-circuits/src/evm_circuit/execution/precompiles/ecrecover.rs#L404-L412).